File: src/main/java/io/humble/video/io/FfmpegIO.java
/*******************************************************************************
 * Copyright (c) 2013, Art Clarke.  All rights reserved.
 *  
 * This file is part of Humble-Video.
 *
 * Humble-Video is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Humble-Video is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Humble-Video.  If not, see <http://www.gnu.org/licenses/>.
 *******************************************************************************/
package io.humble.video.io;


import io.humble.video.io.FfmpegIOHandle;
import io.humble.video.io.IURLProtocolHandlerFactory;
import io.humble.video.io.URLProtocolManager;

/**
 * For Internal Use Only.
 * This is the global (static) object that implements the Humble Video IO library.
 * <p>
 * The following methods are not meant to be generally
 * called by code (although they can be).
 * </p></p>
 * You should never need to call this method.  Calling {@link URLProtocolManager#registerFactory(String, IURLProtocolHandlerFactory)}
 * should cause it to be instantiated.
 *</p></p>
 * They are generally 'thread-aware' but not 'thread-safe', meaning if you use a handle
 * on a thread, you are a responsible for making sure you don't
 * reuse it on other threads.
 * </p></p>
 * They forward into ffmpeg's URL read/write/seek capabilities
 * and allow our test scripts to make sure our URLProtocolHandlers
 * (and FFMPEG's native handlers) work as expected.
 * </p></p>
 * Lastly this class, unlike other classes in the library, does not use SWIG to generate
 * the Java objects, so you need to be careful not to change method names as the corresponding
 * native code relies on specific naming and parameters.
 * </p>
 * @author aclarke
 *
 */
public class FfmpegIO
{

  static
  {
    // this will load the one shared library
    io.humble.ferry.Ferry.init();
    FfmpegIO.init();
    // And force the URLProtocolManager global
    // object to be created.
    URLProtocolManager.init();
  }

  /** Force a load of all native libraries; not normally needed */
  public static void load() {}
  
  /**
   * Internal Only.  Do not use.
   */
  public native static void init();
  
  /**
   * This method is called by URLProtocolManager to register itself as a 
   * protocol manager for different FFMPEG protocols.
   * 
   * @param protocol The protocol
   * @param manager The manager for that protocol.
   */
  static synchronized void registerProtocolHandler(String protocol,
      URLProtocolManager manager)
  {
    native_registerProtocolHandler(protocol, manager);
  }

  public static int url_open(FfmpegIOHandle handle, String filename, int flags)
  {
    return native_url_open(handle, filename, flags);
  }

  public static int url_read(FfmpegIOHandle handle, byte[] buffer, int length)
  {
    return native_url_read(handle, buffer, length);
  }

  public static int url_close(FfmpegIOHandle handle)
  {
    return native_url_close(handle);
  }

  public static int url_write(FfmpegIOHandle handle, byte[] buffer, int length)
  {
    return native_url_write(handle, buffer, length);
  }

  public static long url_seek(FfmpegIOHandle handle, long position, int whence)
  {
    return native_url_seek(handle, position, whence);
  }

  private static native int native_registerProtocolHandler(
      String urlPrefix, URLProtocolManager proto);

  private static native int native_url_open(FfmpegIOHandle handle,
      String filename, int flags);

  private static native int native_url_read(FfmpegIOHandle handle,
      byte[] buffer, int length);

  private static native int native_url_write(FfmpegIOHandle handle,
      byte[] buffer, int length);

  private static native long native_url_seek(FfmpegIOHandle handle,
      long position, int whence);

  private static native int native_url_close(FfmpegIOHandle handle);

}

File: src/main/gnu/src/io/humble/video/customio/FfmpegIO.cpp
/*
 * Copyright (c) 2013-Forward, Andrew "Art" Clarke
 *
 * This file is part of Humble Video.
 * 
 * Humble Video is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Humble Video is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with Humble Video.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Container.h
 *
 *  Created on: Jun 30, 2013
 *      Author: aclarke
 */

#include <stdexcept>

// For EINTR
#include <errno.h>

#include <io/humble/ferry/JNIHelper.h>
#include <io/humble/video/customio/FfmpegIO.h>
#include <io/humble/video/customio/JavaURLProtocolManager.h>

using namespace io::humble::ferry;
using namespace io::humble::video::customio;

// we need to have a non-exception raising  interrupt check here
#undef VS_CHECK_INTERRUPT
#define VS_CHECK_INTERRUPT(retval, __COND__) do { \
    if (__COND__) { \
      JNIHelper* helper = JNIHelper::getHelper(); \
      if (helper && helper->isInterrupted()) \
        (retval) = EINTR > 0 ? -EINTR : EINTR; \
        } \
} while(0)

VS_API_HUMBLE_VIDEO_CUSTOMIO void VS_API_CALL Java_io_humble_video_customio_FfmpegIO_init(JNIEnv *env, jclass)
{
  JavaVM* vm=0;
  if (!io::humble::ferry::JNIHelper::sGetVM()) {
    env->GetJavaVM(&vm);
    io::humble::ferry::JNIHelper::sSetVM(vm);
  }
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jint VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1registerProtocolHandler(
    JNIEnv *jenv, jclass, jstring aProtoName, jobject aProtoMgr)
{
  int retval = -1;
  const char *protoName= NULL;
  try {
    protoName = jenv->GetStringUTFChars(aProtoName, NULL);
    if (protoName != NULL)
    {
      // Yes, we deliberately leak the URLProtocolManager...
      JavaURLProtocolManager::registerProtocol(protoName, aProtoMgr);
      retval = 0;
    }
  }
  catch(std::exception & e)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      if (cls)
        jenv->ThrowNew(cls, e.what());
    }
    retval = -1;
  }
  catch(...)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      if (cls)
        jenv->ThrowNew(cls, "Unhandled and unknown native exception");
    }
    retval = -1;
  }
  if (protoName != NULL) {
    jenv->ReleaseStringUTFChars(aProtoName, protoName);
    protoName = NULL;
  }
  return retval;
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jint VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1url_1open(
    JNIEnv * jenv, jclass, jobject handle, jstring url, jint flags)
{
  URLProtocolHandler*handleVal= NULL;
  const char *nativeURL= NULL;
  jint retval = -1;

  nativeURL = jenv->GetStringUTFChars(url, NULL);
  if (nativeURL != NULL)
  {
    try
    {
      handleVal = URLProtocolManager::findHandler(nativeURL, flags, 0);
      if (!handleVal)
        throw std::runtime_error("could not find protocol manager for url");
      retval = handleVal->url_open(nativeURL, flags);
      VS_CHECK_INTERRUPT(retval, 1);
    }
    catch(std::exception & e)
    {
      // we don't let a native exception override a java exception
      if (!jenv->ExceptionCheck())
      {
        jclass cls=jenv->FindClass("java/lang/RuntimeException");
        jenv->ThrowNew(cls, e.what());
      }
      retval = -1;
    }
    catch(...)
    {
      // we don't let a native exception override a java exception
      if (!jenv->ExceptionCheck())
      {
        jclass cls=jenv->FindClass("java/lang/RuntimeException");
        jenv->ThrowNew(cls, "Unhandled and unknown native exception");
      }
      retval = -1;
    }

    // regardless of what the function returns, set the pointer
    JNIHelper::sSetPointer(handle, handleVal);
  }

  // free the string
  if (nativeURL != NULL)
  {
    jenv->ReleaseStringUTFChars(url, nativeURL);
    nativeURL = NULL;
  }
  return retval;
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jint VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1url_1read(
    JNIEnv *jenv, jclass, jobject handle, jbyteArray javaBuf, jint buflen)
{
  URLProtocolHandler* handleVal= NULL;

  jint retval = -1;

  handleVal = (URLProtocolHandler*)JNIHelper::sGetPointer(handle);

  jbyte *byteArray = jenv->GetByteArrayElements(javaBuf, NULL);
  try
  {
    if (handleVal)
      retval = handleVal->url_read((uint8_t*)byteArray, buflen);
    VS_CHECK_INTERRUPT(retval, retval < 0 || retval < buflen);
  }
  catch(std::exception & e)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, e.what());
    }
    retval = -1;
  }
  catch(...)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, "Unhandled and unknown native exception");
    }
    retval = -1;
  }
  jenv->ReleaseByteArrayElements(javaBuf, byteArray, 0);
  return retval;
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jint VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1url_1write(
    JNIEnv *jenv, jclass, jobject handle, jbyteArray javaBuf, jint buflen)
{
  URLProtocolHandler* handleVal= NULL;

  jint retval = -1;

  handleVal = (URLProtocolHandler*)JNIHelper::sGetPointer(handle);

  jbyte *byteArray = jenv->GetByteArrayElements(javaBuf, NULL);
  try
  {
    if (handleVal)
      retval = handleVal->url_write((uint8_t*)byteArray, buflen);
    VS_CHECK_INTERRUPT(retval, retval < 0 || retval != buflen);
  }
  catch(std::exception & e)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, e.what());
    }
    retval = -1;
  }
  catch(...)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, "Unhandled and unknown native exception");
    }
    retval = -1;
  }
  jenv->ReleaseByteArrayElements(javaBuf, byteArray, 0);
  return retval;
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jlong VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1url_1seek(
    JNIEnv *jenv, jclass, jobject handle, jlong position, jint whence)
{
  URLProtocolHandler* handleVal= NULL;

  jint retval = -1;

  handleVal = (URLProtocolHandler*)JNIHelper::sGetPointer(handle);
  try
  {
    if (handleVal)
      retval = handleVal->url_seek((int64_t)position, whence);
    VS_CHECK_INTERRUPT(retval, 1);
  }
  catch(std::exception & e)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, e.what());
    }
    retval = -1;
  }
  catch(...)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, "Unhandled and unknown native exception");
    }
    retval = -1;
  }
  return retval;
}

VS_API_HUMBLE_VIDEO_CUSTOMIO jint VS_API_CALL Java_io_humble_video_customio_FfmpegIO_native_1url_1close(
    JNIEnv *jenv, jclass, jobject handle)
{
  URLProtocolHandler* handleVal= NULL;

  jint retval = -1;

  handleVal = (URLProtocolHandler*)JNIHelper::sGetPointer(handle);
  try
  {
    if (handleVal)
      retval = handleVal->url_close();
    VS_CHECK_INTERRUPT(retval, 1);
  }
  catch(std::exception & e)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, e.what());
    }
    retval = -1;
  }
  catch(...)
  {
    // we don't let a native exception override a java exception
    if (!jenv->ExceptionCheck())
    {
      jclass cls=jenv->FindClass("java/lang/RuntimeException");
      jenv->ThrowNew(cls, "Unhandled and unknown native exception");
    }
    retval = -1;
  }
  return retval;
}
